rust/src/common/sync.rs

```diff
@@ -0,0 +1,184 @@
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::{Duration, SystemTime},
+};
+use tokio::{
+    stream::StreamExt,
+    sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender},
+};
+
+pub enum SyncTasks {
+    Enter,
+    Reset,
+    Exit,
+}
+
+#[async_trait]
+pub trait ProcessSync {
+    async fn enter_sync(&self);
+    async fn reset(&self);
+    async fn exit_sync(&self);
+}
+
+#[async_trait]
+pub trait SendReset {
+    async fn reset(&mut self, ctx: tarpc::context::Context) -> std::io::Result<()>;
+}
+
+#[derive(Debug)]
+pub enum SyncRequest {
+    // The RPC server received a sync request from the aggregator.
+    ExternalRequest,
+    // The RPC client lost the connection to the aggregator RPC server.
+    RPCClientDisconnect,
+}
+
+pub struct SyncService<S, R>
+where
+    S: ProcessSync + Clone + Send + Sync + Unpin + 'static,
+    R: SendReset + Clone + Send + Sync + Unpin + 'static,
+{
+    service_handle: S,
+    rpc_client: R,
+    state_handle: StateHandle,
+    sync_requests: SyncRequests,
+}
+
+impl<S, R> SyncService<S, R>
+where
+    S: ProcessSync + Clone + Send + Sync + Unpin + 'static,
+    R: SendReset + Clone + Send + Sync + Unpin + 'static,
+{
+    pub fn new(
+        service_handle: S,
+        rpc_client: R,
+        state_handle: StateHandle,
+        sync_requests: SyncRequests,
+    ) -> Self {
+        Self {
+            service_handle,
+            rpc_client,
+            state_handle,
+            sync_requests,
+        }
+    }
+
+    /// Handle the incoming requests.
+    fn poll_requests(&mut self, cx: &mut Context) -> Poll<()> {
+        trace!("polling requests");
+        loop {
+            match ready!(Pin::new(&mut self.sync_requests).poll_next(cx)) {
+                Some(request) => {
+                    self.handle_request(request);
+                }
+                None => return Poll::Ready(()),
+            }
+        }
+    }
+
+    /// Handle a request
+    fn handle_request(&mut self, request: SyncRequest) {
+        match request {
+            SyncRequest::RPCClientDisconnect => self.handle_rpc_client_disconnect_request(),
+            SyncRequest::ExternalRequest => self.handle_external_request(),
+        }
+    }
+
+    /// Handle a rpc client disconnect request
+    fn handle_rpc_client_disconnect_request(&mut self) {
+        debug!("handle rpc_client_disconnect request");
+        let service_handle = self.service_handle.clone();
+        let mut rpc_client = self.rpc_client.clone();
+        let _ = tokio::spawn(async move {
+            service_handle.enter_sync().await;
+            service_handle.reset().await;
+
+            let mut context = tarpc::context::current();
+            context.deadline = SystemTime::now() + Duration::from_secs(120);
+            rpc_client
+                .reset(context)
+                .await
+                .expect("Timeout. Could not connect to the external service");
```

Comment:
Is it reasonable to panic here? If out client disconnected, it's possible that it wont be able to reconnect within 2 minutes.